### ISSUE TITLE
mongoDB

### ISSUE THREAD

**Issue by jamesaxl** [loentar/ngrest-db#6]
what do you think about adding mongodb, i did some test and it works.

**Comment by loentar** [comment]
Hmm, why not? Can I take a look on your implementation (pool request, please)?

**Comment by jamesaxl** [comment]
of Course, give me time please.

**Comment by jamesaxl** [comment]
could you show me the code that use for getting  // *table: name and also
field // *unique: true ?
thankyou,

**Comment by loentar** [comment]
Entity interface store such information. Implementation is code-generated at build time for every DB entity.

To get table name you can use `entity.getTableName()`, Example:
https://github.com/loentar/ngrest-db/blob/master/drivers/sqlite/src/SQLiteDb.cpp#L272


To get access to field info `Field` interface is used, It's code-generated as well.
See `field.isUnique`:
https://github.com/loentar/ngrest-db/blob/master/drivers/sqlite/src/SQLiteDb.cpp#L252


Please also look to the whole `getCreateTableQuery` method, It uses all the meta-info passed from code-generator: https://github.com/loentar/ngrest-db/blob/master/drivers/sqlite/src/SQLiteDb.cpp#L237

Also please look to Entity https://github.com/loentar/ngrest-db/blob/master/common/src/Entity.h
and Field https://github.com/loentar/ngrest-db/blob/master/common/src/Field.h



I think the easiest method to develop a new driver is to copy any existing, rename and modify it's implementation.

**Comment by jamesaxl** [comment]
yes that it is right, but Mongodb it a little different, it is nosql. I
will implement anyway on the core of your repo and see how it will work.
Thank you.

**Comment by jamesaxl** [comment]
Could you explain to me this code,

class NotesDb: public ::ngrest::DbManager< ::ngrest::PostgresDb > //is
::ngrest::DbManager map?
{
public:
    static NotesDb& inst()
    {
        static NotesDb instance; // ths instance is a DbManager instance or
PostgresDb instance?
        return instance;
    }

thank you

**Comment by loentar** [comment]
DbManager is a helper template class to store instance of db driver and to provide access to tables. You have to pass db driver class as template parameter to use it. All db drivers are inherited from `Db` class and provide the same interface. So if you implement a new db driver with the same interface, you just have to use it like `DbManager<MongoDb>` without any rewriting in ngrest-db code.

The `static NotesDb& inst()` is a singleton accessor with lazy creation. It's to have just one instance of db in memory. You have to call it like that: `NotesDb::inst().db()` - to get raw db instance, or `NotesDb::inst().getTable<Notes>()` - to get notes table.

**Comment by jamesaxl** [comment]
Thank you very much, I understand it very, the problem now is : MongoDB it
is not a sql database manager it use bson for creating collection and
document, i can not use DbManager, and if I want use it I should re-write
it, and it is not good :), cause it works wotk this three database-manager
very well
, I am going to find how to implement on ngrest-db.

**Comment by loentar** [comment]
I'm afraid `DbManager` needs only few lines to change.

If you look into the `Table` class, you can see it contain many SQL code.. To add NoSQL support need to generalize `Table` into abstract class and derive implementation classes like `SqlTable` and `NoSqlTable`. Also Db will be split to `SqlDb`(which used with `SqlTable`) and `NoSqlDb` (for `NoSqlTable`).

So many changes.. I can try to do this work, but this will be not quick and if you can provide the MongoDB driver. Can you develop the NoSQL driver with methods which cover the Table functionality (select, insert, deleteWhere, etc.)? I will generalize `Db` and `Table` classes then.

**Comment by jamesaxl** [comment]
Hi,
Do you join some channels in IRC? if yes could you tell me which one.

**Comment by loentar** [comment]
Sorry, I don't use IRC.
Feel free to contact me via Email ([the same username as on github]@gmail.com) and Telegram ([the same username as on github]).

Also I created chat on gitter.im/ngrest

**Comment by jamesaxl** [comment]
Hello, Could i share your application on my repo
https://fossil.falseking.site .

**Comment by loentar** [comment]
I'm not sure what you mean under "application", but if you mean this project or ngrest itself, you free to share it, it's under Apache 2 license.
